### PR TITLE
DBZ-443 db-history recovery mode

### PR DESCRIPTION
**Pull request by pgoranss**
Adding db-history recovery mode: schema_only_recovery

See original PR for add'l comments https://github.com/debezium/debezium/pull/375 (accidentally closed) 

Description of changes:
If the db-history topic becomes corrupted and causes the connector to get stuck, this snapshot mode will re-snapshot the schema just like schema_only, but will continue reading the binlog from it's existing position rather than the current position.

This is different than schema-only as it allows us to gracefully recover the records during the outage gap, assuming the schema has not changed during that time. If the schema has changed, it may result in some corrupted events during the gap, however this is probably more desirably than lost events anyway.

### CODE REVIEW

**Comment by pgoranss** [comment]
@jpechane I've incorporated your comments from https://github.com/debezium/debezium/pull/375 . I like the idea of adding a general recovery mode configuration. That maybe should be a separate enhancement request? Once that is complete this can easily be moved over. In the meantime this can be an undocumented snapshot mode.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java [hunk 1]
```diff
@@ -666,29 +667,39 @@ protected void execute() {
...
-                    logger.info("\t using binlog '{}' at position '{}' and gtid '{}'", binlogFilename, binlogPosition,
-                                gtidSet);
+        if (context.isSchemaOnlyRecoverySnapshot()) {
+            // We are in schema only recovery mode, use the existing binlog position            
+            if (source.binlogFilename() == null || source.binlogFilename().isEmpty()) {
+                // would like to also verify binlog position exists, but it defaults to 0 which is technically valid
+                throw new IllegalStateException("Could not find existing binlog information while attempting schema only recovery snapshot");
+            }
+            source.startSnapshot();
+        }
```
Comment (jpechane) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java]: Again formatting

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 2]
```diff
@@ -257,7 +257,22 @@ private Long getEndOffsetOfDbHistoryTopic(Long previousEndOffset, KafkaConsumer<
 
         return endOffset;
     }
-
+    
+    @Override
+    public boolean exists() {
+        boolean exists = false;
+        try (KafkaConsumer<String, String> historyConsumer = new KafkaConsumer<>(consumerConfig.asProperties());) {
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Just a question - will this work even if the topic is missing? I think we need to guard against both cases. Sorry if i was not clear.

Comment (pgoranss) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Yes, I tested both cases

**Comment by pgoranss** [comment]
I have added some unit tests and fixed the formatting, per review comments

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java [hunk 3]
```diff
@@ -158,6 +158,7 @@ protected Object readField(ResultSet rs, int fieldNo, Column actualColumn) throw
      */
     private Object readTimeField(ResultSet rs, int fieldNo) throws SQLException {
         Blob b = rs.getBlob(fieldNo);
+        if (b == null) return null; // Don't continue parsing time field if it is null
```
[the remark this answers is not in the document]
Comment (pgoranss) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java]: This is already in master branch:
https://github.com/debezium/debezium/blob/master/debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java#L161

I'm not sure why it's showing as a new line, I will try to resolve this

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 4]
```diff
@@ -257,7 +257,22 @@ private Long getEndOffsetOfDbHistoryTopic(Long previousEndOffset, KafkaConsumer<
...
+    public boolean exists() {
+        boolean exists = false;
+        try (KafkaConsumer<String, String> historyConsumer = new KafkaConsumer<>(consumerConfig.asProperties());) {                        
+            Map<TopicPartition, Long> beginningOffsets = historyConsumer.beginningOffsets(Collections.singleton(new TopicPartition(topicName, PARTITION)));
+            Map<TopicPartition, Long> endOffsets = historyConsumer.endOffsets(Collections.singleton(new TopicPartition(topicName, PARTITION)));            
+            
+            Long beginOffset = beginningOffsets.entrySet().iterator().next().getValue();
+            Long endOffset = endOffsets.entrySet().iterator().next().getValue();            
+            
+            exists = (endOffset - beginOffset) != 0;
```
[the remark this answers is not in the document]
Comment (pgoranss) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Will adjust this

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 5]
```diff
@@ -257,7 +257,22 @@ private Long getEndOffsetOfDbHistoryTopic(Long previousEndOffset, KafkaConsumer<
 
         return endOffset;
     }
-
+    
+    @Override
+    public boolean exists() {
+        boolean exists = false;
+        try (KafkaConsumer<String, String> historyConsumer = new KafkaConsumer<>(consumerConfig.asProperties());) {                        
+            Map<TopicPartition, Long> beginningOffsets = historyConsumer.beginningOffsets(Collections.singleton(new TopicPartition(topicName, PARTITION)));
```
[the remark this answers is not in the document]
Comment (pgoranss) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: I think that would be a configuration issue. I originally had exists() simply get the list of topics and look for the name as in the code below, but @jpechane pointed out we need to check for an empty topic as well as non-existing. Do you think I should add that topic name check back in as a pre-check?
```
// Precheck
exists = historyConsumer.listTopics().keySet().contains(topicName);
```

[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Yes, I think that'd be a good thing to do.

**Comment by gunnarmorling** [comment]
Hey @pgoranss, we've planned to do the 0.7.2 Wednesday this week. Do you think you could address the last few comments until then? Thanks a lot!

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pgoranss) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java]: I just had to merge debezium/master into the branch again to make this diff go away.

**Comment by pgoranss** [comment]
Hey @gunnarmorling, I have checked in changes to address those last few comments, thanks

**Comment by gunnarmorling** [comment]
@pgoranss Awesome, I'll check it out and merge asap.

In the meantime, could you add the new snapshotting mode to the [reference docs](https://github.com/debezium/debezium.github.io/blob/develop/docs/connectors/mysql.asciidoc), too? The snapshotting modes are described in the text and in the table of configuration options towards the end. Shouldn't be too much work to add the new mode there, too.

Thanks!

**Comment by pgoranss** [comment]
@gunnarmorling I have added some documentation: https://github.com/debezium/debezium.github.io/pull/110 

It is articulated from the point of view of using a kafka database history topic.

**Comment by gunnarmorling** [comment]
Rebased, squashed everything into a single commit and applied. Thanks a lot, @pgoranss!

**Pull request closed without merging**